x-pack/plugin/fleet/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/fleet/40_secrets_get.yml

```diff
@@ -3,6 +3,10 @@
       fleet.post_secret:
         body: '{"value": "test secret"}'
   - set:  { id: id }
+  # search node needs to be available for fleet.get_secret to work
+  - do:
+      cluster.health:
+        wait_for_status: green
```

Comment:
Thanks @bcully for figuring this out!

Could we instead use 

```
  - do:
      cluster.health:
        wait_for_no_initializing_shards: true
```

as done in https://github.com/elastic/elasticsearch/pull/114641 ? It was a similar situation.

Reply:
sounds good, thanks for the review and the additional context

Reply:
that didn't seem to work for me, somewhat surprisingly:

```
    [2024-11-21T08:59:41,869][INFO ][c.e.e.s.StatelessIndexEventListener] [index-rdEfLBd] [.fleet-secrets-7][0] with UUID [YMfTE28qTpWR_z8rqel1uw] bootstrapping [INDEX_ONLY] shard on primary term [1] with empty commit from object store (new shard recovery)
    [2024-11-21T08:59:41,979][INFO ][o.e.i.e.Engine           ] [index-rdEfLBd] [.fleet-secrets-7][0] Completed force merge for shard [.fleet-secrets-7][0] without flushing
    [2024-11-21T08:59:42,125][INFO ][o.e.x.f.FleetRestIT      ] [test] Dump test yaml [/Users/brendan/dev/elasticsearch-serverless/qa/stateful/x-pack/plugin/fleet/qa/rest/build/yamlTests/rest-api-spec/test/fleet/40_secrets_get.yml] on failure:
    "Get Fleet Secret":
      - do:
          fleet.post_secret:
            body: '{"value": "test secret"}'
      - set:  { id: id }
      # search node needs to be available for fleet.get_secret to work
      - do:
          cluster.health:
            wait_for_no_initializing_shards: true
      - do:
          fleet.get_secret:
            id: $id
      - match: { id: $id }
      - match: { value: "test secret" }
      - do:
          fleet.delete_secret:
            id: $id

    ---
    "Get non existent Fleet Secret":
      - do:
          catch: "missing"
          fleet.get_secret:
            id: "foo"

    ---
    "Get secret fails for unprivileged user":
      - skip:
          features: headers
      - do:
          catch: forbidden
          headers: { Authorization: "Basic ZmxlZXRfdW5wcml2aWxlZ2VkX3NlY3JldHM6cGFzc3dvcmQ=" } # base64 encoded fleet_unprivileged_secrets:password
          fleet.get_secret:
            id: "foo"

    [2024-11-21T08:59:42,126][INFO ][o.e.x.f.FleetRestIT      ] [test] Stash dump on test failure [{
      "stash" : {
        "id" : "WYI8T5MBg_EpYGiZcnev",
        "body" : {
          "error" : {
            "root_cause" : [
              {
                "type" : "no_shard_available_action_exception",
                "reason" : "No shard available for [get [.fleet-secrets][WYI8T5MBg_EpYGiZcnev]: routing [null]]",
                "stack_trace" : "org.elasticsearch.action.NoShardAvailableActionException: No shard available for [get [.fleet-secrets][WYI8T5MBg_EpYGiZcnev]: routing [null]]
    "
              }
            ],
            "type" : "no_shard_available_action_exception",
            "reason" : "No shard available for [get [.fleet-secrets][WYI8T5MBg_EpYGiZcnev]: routing [null]]",
            "stack_trace" : "org.elasticsearch.action.NoShardAvailableActionException: No shard available for [get [.fleet-secrets][WYI8T5MBg_EpYGiZcnev]: routing [null]]
    "
          },
          "status" : 503
        }
      }
    }]
    [2024-11-21T08:59:42,138][INFO ][o.e.x.f.FleetRestIT      ] [test] Waiting for all cluster updates up to this moment to be processed
    [2024-11-21T08:59:42,627][INFO ][o.e.c.u.FeatureFlag      ] [search-fbQXQPQ] The current build is a snapshot, feature flag [zstd_stored_fields] is enabled
    [2024-11-21T08:59:42,662][INFO ][o.e.x.m.MachineLearning  ] [index-rdEfLBd] Starting machine learning feature reset
    [2024-11-21T08:59:42,669][INFO ][c.e.e.s.StatelessIndexEventListener] [search-fbQXQPQ] [.fleet-secrets-7][0] with UUID [YMfTE28qTpWR_z8rqel1uw] bootstrapping [SEARCH_ONLY] shard on primary term [1] with [stateless_commit_3][1][3] and latest uploaded [term=1, gen=3] from indexing shard (peer recovery from index-rdEfLBd)
    [2024-11-21T08:59:42,700][INFO ][o.e.c.r.a.AllocationService] [index-rdEfLBd] current.health="GREEN" message="Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[.fleet-secrets-7][0]]])." previous.health="YELLOW" reason="shards started [[.fleet-secrets-7][0]]"
```

Reply:
@bcully weird :/ Seems like it tries the GET before the shard is even allocated? Did you check the server logs for any errors?

But OK, if the ensureGreen works, it's fine as well. Not sure we need to spend too much time/effort since it's a test issue.

Feel free to make the CI green with the PR and I'll approve.

Reply:
> Seems like it tries the GET before the shard is even allocated?

That's what it seems like to me, like a search shard is allocated separately for this `.fleet-secrets` index. I don't understand the machinery that creates that index at this point. I'll check out the logs and dig a little bit before I merge, but won't spend too much time on it.

Reply:
It's also a little surprising to me that the cluster goes green after this test step with
```
    [2024-11-21T19:05:44,079][INFO ][o.e.c.r.a.AllocationService] [index-ldEmIVk] current.health="GREEN" message="Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[.fleet-secrets-7][0]]])." previous.health="YELLOW" reason="shards started [[.fleet-secrets-7][0]]"
```
as if it was yellow because it was waiting for the search shard, which to me would imply that `wait_for_no_initializing_shards` should have seen it (going by the description).

Reply:
Nothing interesting in the logs for any of the three nodes (index/search/ml). Does stateless do anything differently around replica allocation on account of having a separate tier for search nodes perhaps?

Reply:
So assuming the index settings have 1 primary and 1 replica configured, I would have expected the wait_for_no_initializing_shards to work. Maybe there's some small windows where the search shard is not yet allocated before it appears as initializing? But then I wonder why the other PR I mentioned works ... maybe it does a bit more stuff before actually searching.

The other case could be if the test for whatever reason (didn't look if there's any custom code in the suite) might start with 1 primary and 0 replicas, and then increases the replicas to 1. Improbable I think.

In any case, I'm happy if green works, I'll approve. But if you'd like to spend more time so we understand why (totally optional), feel free to "debug" it, i.e., observe or log the allocation changes while the test is ongoing to see the transitions.

Reply:
I didn't see anything interesting at all in the `FleetRestIT` setup code, FWIW. `.fleet-secrets` looks like a standard system index from what I've seen (except that it's installed by a plugin).

Reply:
One thing about this index, it is an alias for an underlying `.fleet-secrets-7` index.

Reply:
So my guess is, the put creates the `.fleet-secrets-7` shard on demand. When it does, the cluster state becomes yellow because its associated search replica isn't assigned yet. But since it isn't assigned, it also isn't initializing, which is why `wait_for_no_uninitialized_shards` works.

We don't seem to have a `wait_for_no_unassigned_shards`, I guess because the use case overlaps so much with waiting for green. I think waiting for green (or all shards to be active) is the right move here based on that. I can
update the comment in the test with more info though.

Reply:
> When it does, the cluster state becomes yellow because its associated search replica isn't assigned yet. But since it isn't assigned, it also isn't initializing, which is why wait_for_no_uninitialized_shards works.

Yup, that makes sense to me!

Thanks for confirming